xs/src/libslic3r/PerimeterGenerator.cpp

```diff
@@ -463,6 +492,541 @@ PerimeterGenerator::_traverse_loops(const PerimeterGeneratorLoops &loops,
+                }
+                if (length_poly_2 > length_trim_2) {
+                    inner_end->polyline.clip_end(length_trim_2);
+                } else {
+                    inner_end->polyline.points.erase(
+                        inner_end->polyline.points.begin() + 1,
+                        inner_end->polyline.points.end());
+                }
+            }
+
+            //last check to see if we need a reverse
+            {
+                Line l1(outer_start->polyline.points.back(), inner_start->polyline.points.front());
+                Line l2(inner_end->polyline.points.back(), outer_end->polyline.points.front());
+                Point p_inter(0, 0);
+                bool is_interect = l1.intersection(l2, &p_inter);
+                if (is_interect && p_inter.distance_to(l1) < SCALED_EPSILON && p_inter.distance_to(l2) < SCALED_EPSILON) {
+                    //intersection! need to reverse!
+                    std::reverse(my_loop.paths.begin() + nearest.idx_polyline_outter + 1, my_loop.paths.begin() + nearest.idx_polyline_outter + child_paths_size + 1);
+                    for (size_t idx = nearest.idx_polyline_outter + 1; idx < nearest.idx_polyline_outter + child_paths_size + 1; idx++) {
+                        my_loop.paths[idx].reverse();
+                    }
+                    outer_start = &my_loop.paths[nearest.idx_polyline_outter];
+                    inner_start = &my_loop.paths[nearest.idx_polyline_outter + 1];
+                    inner_end = &my_loop.paths[nearest.idx_polyline_outter + child_paths_size];
+                    outer_end = &my_loop.paths[nearest.idx_polyline_outter + child_paths_size + 1];
+                }
+
+            }
+
+            //now add extrusionPAths to connect the two loops
+            ExtrusionPaths travel_path_begin;
+            ExtrusionPaths travel_path_end;
+            double dist_travel = outer_start->polyline.points.back().distance_to(inner_start->polyline.points.front());
+            if (dist_travel > max_width_extrusion*1.5 && this->config->fill_density.value > 0) {
+                travel_path_begin.emplace_back(ExtrusionRole::erPerimeter, outer_start->mm3_per_mm, outer_start->width, outer_start->height);
+                travel_path_begin.emplace_back(ExtrusionRole::erNone, 0, outer_start->width, outer_start->height);
+                travel_path_begin.emplace_back(ExtrusionRole::erPerimeter, outer_start->mm3_per_mm, outer_start->width, outer_start->height);
+                //travel_path_begin[0].extruder_id = -1;
+                //travel_path_begin[1].extruder_id = -1;
+                //travel_path_begin[2].extruder_id = -1;
+                Line line(outer_start->polyline.points.back(), inner_start->polyline.points.front());
+                Point p_dist_cut_extrude = (line.b - line.a);
+                p_dist_cut_extrude.x = (coord_t)(p_dist_cut_extrude.x * ((double)max_width_extrusion) / (line.length() * 2));
```

Comment:
scale()

Also - longer line is cut by smaller amount, is that desirable?

Reply:
Sorry, the code is fine ... 
But the normalization is very well hidden .. 
It may be better to switch to coordf_t, normalize and scale